### ISSUE TITLE
API docs does not list RNNs

### ISSUE THREAD

**Issue by AvantiShri** [tensorflow/tensorflow#7]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/g3doc/api_docs/python/nn.md

(It has convolutional layers listed, for instance, but does not show the RNNs. Actually, I don't quite see anything corresponding to a fully-connected layer either)

**Comment by manning** [comment]
You should look at the RNN tutorial: http://tensorflow.org/tutorials/recurrent/index.md .

**Comment by AvantiShri** [comment]
(Thanks, I have seen the tutorial but it is not a substitute for an API; I filed the issue after consulting with a friend at Google Brain)

**Comment by vrv** [comment]
Hi Avanti -- internally we've been working on iterating the API for RNNs, and we were happy enough with the current API to use it in the tutorial, but we're making sure it's solid before promoting it to the public API, since we'd then have to support it indefinitely.  (Anything not in the public API is a work-in-progress :)

We'll keep this bug open in the meantime, and for now you can look at the source code documentation if you're interested in playing around: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/rnn.py#L9

**Comment by AvantiShri** [comment]
Ah, got it, thanks for the explanation.

**Comment by zer0n** [comment]
The white paper of TensorFlow mentions looping control within the graph. Is it already available? If so, are there examples to show how it can be done?

The [RNN example] (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/rnn.py#L86) has a Python loop. Will TensorFlow treat that as a symbolic loop and compile it?

Also, the explanation of `sequence_length` [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/rnn.py#L30) isn't clear to me. What does it mean by **dynamic** calculations? When `t` is past `max_sequence_length`, can it just break from the loop instead of continuing with `zeros` state? Returning `zeros` state is different from returning the state at `max_sequence_length`, isn't it?

**Comment by vrv** [comment]
On your first question, see https://github.com/tensorflow/tensorflow/issues/208.

On your second question: the core TF engine currently only sees the GraphDef produced by python, so the RNN example is an unrolled one today.

I'm not super familiar with that RNN example -- @lukaszkaiser or @ludimagister might know better.

**Comment by ludimagister** [comment]
I zer0n,

the current RNN is statically unrolled, there is no (not yet) dynamic unrolling based on the length of the sequence. Th dynamic calculation means the graph is unrolled up to the max_sequence_length, but if a sequence_length is provided the calculations on the unrolled graph are cut short once the sequence_length is reached, using a conditional op. Depending on the application this may result in shorter processing time.

**Comment by ebrevdo** [comment]
Yes, to add to what @ludimagister says: the conditional op will plug in zeros to the output & state past max(sequence_length), thus reducing the total amount of computation (if not memory).

I may actually modify this so that instead of plugging in zeros to the state, it just copies the state.  This way the final state will represent the "final" state at max(sequence_length).  However, I'm undecided on this.  If you want the final state at time sequence_length, you can concat the state vectors and use transpose() followed by gather() with sequence_length in order to pull out the states you care about.  That's probably what you would want to do, in fact, because if you have batch_size = 2 and sequence_length = [1, 2], then for the first minibatch entry, the state at max(sequence_length) will not equal the state at sequence_length[0].

An alternative solution is to right-align your inputs so that they always "end" on the final time step.  This breaks down the dynamic calculation performed when you pass sequence_length (because it assumes left-aligned inputs).  I may extend this by adding a bool flag like "right_aligned" to the rnn call, which assumes that calculation starts at len(inputs) - max(sequence_length), and copies the initial state through appropriately.  But that doesn't exist now.

**Comment by zer0n** [comment]
Thanks @vrv, @ludimagister, and @ebrevdo for the answers. However, Some details still confuse me.

1. @ludimagister, the code doesn't seem to statically unroll. It has [a loop which depends on the length of the inputs](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/rnn/rnn.py#L86). Plus, `max_sequence_length` is not a const; instead it's just the scalar of the `sequence_length` parameter which can be `None`. So, by default, `sequence_length` is `None` and the unrolling is not truncated. Correct me if I misread the code.

2. @ebrevdo I understand the computational saving motivation. However, returning zeros is logically very different from returning the state at `sequence_length` (if provided). The former is just wrong. Again, please correct if I misread the code.

Thanks.

**Comment by ebrevdo** [comment]
@zer0n It depends on your task.

Returning zeros is fine if you only care about outputs (i.e., you're not hooking up to a decoder); and your loss function knows to ignore outputs past the sequence_length.

Returning the state from the end of the last time step might also be considered "wrong", but will generally always happen if you have inputs of different lengths (and aren't performing dynamic computation).  This is a typical approach to performing RNN with minibatches.  For this reason when performing encoding/decoding, people usually right-align with left-side padding instead, so the last input of any example always corresponds to the very last state.  This seems like the cleanest solution for now.

Anyway, this part of the API may change; not sure yet the best approach.

**Comment by zer0n** [comment]
OK, I did miss the line `outputs.append(output)`. I originally thought that it returned the final state, not a sequence of states.

Anyway, this implementation still looks weird (I'm aware it's changing so I'm only discussing the current state). Usually, for truncated BPTT implementation, people pad `eos` for short sentences and truncate the sentences if the lengths are larger than `max_length`. This enables static unrolling and efficient mini-batching.

The RNN example seems doing the reverse. What I see is that it's doing **dynamic** unrolling (i.e. with dynamic output size), but padding zeros to the outputs past `max_length`.

**Comment by vrv** [comment]
(This discussion is probably better off had on the discussion mailing list, rather than this bug about documentation)

**Comment by martinwicke** [comment]
@webmaven: We are going to be using semver, and we will publish exactly what we mean by that too. We'll try reasonably hard to maintain API stability even before 1.0.0, especially in the parts of the API that are official. For now, we have decided that something becomes "official" when it shows up on the API docs. You'll notice that many other functions are documented but not included in the docs, that means their interface is still in flux.

I'm closing this bug for now.

**Comment by webmaven** [comment]
@martinwicke:
> We'll try reasonably hard to maintain API stability even before 1.0.0, especially in the parts of the API that are official. For now, we have decided that something becomes "official" when it shows up on the API docs. 

Hmm. OK, but wouldn't putting APIs in the docs and marking them as 'draft' or 'unofficial' get you valuable external feedback while you're still iterating on an API? Or is external feedback only wanted/needed *after* an API becomes 'official'?

**Comment by vrv** [comment]
Yeah, we're definitely considering adding something like this since it helps to get community feedback on some experimental APIs before they are fully ready.

**Comment by webmaven** [comment]
@vrv and @martinwicke, so getting back to the original topic, is the RNN API currently a good candidate for such 'draft' treatment, and does that mean this issue on documentation should be re-opened?

**Comment by ebrevdo** [comment]
We are reconsidering the api, deciding if we should make the current
rnncell implementations stateful, or whether to store some of the non
variable state in graph collections.  After that decision is made, we will
likely add documentation.

**Comment by webmaven** [comment]
@ebrevdo what are the considerations surrounding this decision between stateful rnncell implementations vs. storing the state in the graph collections?

**Comment by ebrevdo** [comment]
The main consideration is in how calculations common across time are
cached.  Simpler caching is traded off against having RNNCell be a pure
logic object (non-stateful).

For example, if you access two Variables at each time step and then concat
them, using the result in your __call__ calculation, then this is something
that should be cached beforehand because it creates redundant computation.
The two approaches to caching are:

1. RNNCell is stateful: create and cache this Tensor inside the RNNCell
object
2. RNNCell is non-stateful: __call__ looks for the cached Tensor inside a
graph collection; if it doesn't exist, it creates it (similar to using
get_variable).

With a stateful RNNCell, Variables are created when the RNNCell is created;
and so that variable scope is used.  With a non-stateful RNNCell, Variables
are created / accessed during __call__ and the variable scope used is
whatever it was when you ran rnn() or bidirectional_rnn() or whatever.

Because of this, moving from non-stateful RNNCell to a stateful one (and
modifying the associated implementations of LSTM, GRU, etc cells) would be
a breaking change.

I personally prefer stateful objects, because it's easier to understand and
debug them.  But there are arguments in both directions that have to be
considered.

**Comment by webmaven** [comment]
@ebrevdo:
> [snip explanation]
> But there are arguments in both directions that have to be considered.

What are those arguments?

**Comment by ebrevdo** [comment]
For example:

Moving to stateful objects now would break a bunch of external dependencies.  Of course, this is an undocumented API and therefore folks should expect it to break in their projects.  However, I'm afraid of breaking external projects in subtle ways that don't emit errors.  This indeed may happen with this change.  Especially for those who depth-stack RNNs on top of each other using the same instance.

In addition, there are those who argue that the RNNCell should continue to be a purely logical object with no state, so you can reuse the same instance of RNNCell across multiple RNNs without fear of reusing the same variable in multiple places (though get_variable's checks for over-sharing may ameliorate this somewhat).

**Comment by webmaven** [comment]
So, you *can't* actually reuse the same RNNCell across multiple RNNs in either case?

**Comment by ebrevdo** [comment]
Currently you can.  You can also use it with a shared name scope to tie the
parameters across multiple rnns.

**Comment by webmaven** [comment]
OK, so you *can* currently reuse an instance of RNNCell across multiple RNNs without worrying about reusing the same variable in multiple places.

Does using a shared name scope with a single RNNCell instance across multiple RNNs gain you anything? Or is that really for reusing parameters across multiple RNNCell instances in multiple RNNs?

**Comment by ebrevdo** [comment]
Right. It gains you the ability to tie parameters not only within one lstm,
but also across multiple lstms.
